File: Data Structures & Algorithms/coin-change-ii/submission-1.cpp
class Solution {
public:
    int solve(int amount, vector<int>& coins, int i, vector<vector<int>>& dp) {
        if(amount < 0 || i>=coins.size()) return 0;
        if(amount == 0) return 1;
        if(dp[i][amount] != -1) return dp[i][amount];
        return dp[i][amount] = solve(amount, coins, i+1, dp) + solve(amount-coins[i], coins, i, dp);
    }

    int change(int amount, vector<int>& coins) {
        vector<vector<int>> dp(coins.size()+1, vector<int>(amount+1, -1));
        return solve(amount, coins, 0, dp);
    }
};

File: Data Structures & Algorithms/coin-change/submission-1.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        n = len(coins)
        dp = [[-1] * (amount + 1) for _ in range(n + 1)]
        def count (i: int, rem: float) -> int or float('inf'):
            if rem == 0:
                return 0
            if i >= n or rem < 0:
                return float('inf')
            if dp[i][rem] != -1:
                return dp[i][rem]
            take = count(i, rem - coins[i]) + 1
            skip = count(i+1, rem)
            dp[i][rem] = min(take, skip)
            return dp[i][rem]
        res = count(0, amount)
        return res if res != float('inf') else -1